src/qilib/data_set/data_set.py

```diff
@@ -136,17 +138,32 @@ def sync_from_storage(self, timeout: int) -> None:
             self._storage_reader.sync_from_storage(timeout)
 
     def save_to_storage(self) -> None:
-        """ Not implemented yet. """
-        raise NotImplementedError("save_to_storage has not been implemented.")
+        """ Save DataSet to underlying storage."""
+        for metadata in ['name', 'time_stamp', 'user_data', 'default_array_name']:
+            self._add_metadata_to_storage(metadata, getattr(self, metadata))
+        for array in self._set_arrays:
+            self._add_array_to_storage(array)
+        for array in self._data_arrays.values():
+            self._add_array_to_storage(array)
 
     def finalize(self) -> None:
         """ Indicates to DataSetIOWriters that no more data will be written."""
+        self.save_to_storage()
         for storage in self._storage_writer:
             storage.finalize()
 
+    def add_storage_writer(self, storage_writer: Any) -> None:
+        """ Adda new DataSetIOStorageWriter to the Dataset
+
+        Args:
+            storage_writer: A new initialized storage writer.
+
+        """
+        self._storage_writer.append(storage_writer)
+
     @property
     def storage(self) -> Any:
-        return self._storage_writer
+        return self._storage_writer or self._storage_reader
```

Comment:
That's true, this should be fixed. I thought that though it is strange to have both, it should not necessarily be illegal. But maybe it should in order to avoid some unforeseen complications.